File: src/data/courses/dsa/trees/traversal-block/preorder/code/python/preorder.py
# ============================================================
# Binary Tree Preorder Traversal
# ============================================================

from typing import List

class Solution1:
    def preorderTraversal(self, root) -> List[int]:
        result = []
        def dfs(node):
            if not node: return
            result.append(node.val)
            dfs(node.left)
            dfs(node.right)
        dfs(root)
        return result

class Solution2:
    def preorderTraversal(self, root) -> List[int]:
        if not root: return []
        result, stack = [], [root]
        while stack:
            node = stack.pop()
            result.append(node.val)       # Process root
            if node.right: stack.append(node.right)  # Right first
            if node.left: stack.append(node.left)    # Left next
        return result

File: src/data/courses/dsa/trees/traversal-block/preorder/code/java/Preorder.java
// ============================================================
// Binary Tree Preorder Traversal
// ============================================================

import java.util.*;

class Solution1 {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        dfs(root, res); return res;
    }
    void dfs(TreeNode n, List<Integer> r) {
        if (n==null) return;
        r.add(n.val); dfs(n.left, r); dfs(n.right, r);
    }
}

class Solution2 {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if (root==null) return res;
        Deque<TreeNode> st = new ArrayDeque<>();
        st.push(root);
        while (!st.isEmpty()) {
            TreeNode node = st.pop();
            res.add(node.val);
            if (node.right!=null) st.push(node.right);
            if (node.left!=null) st.push(node.left);
        }
        return res;
    }
}
